lib/purl/generator.ex

```diff
@@ -0,0 +1,84 @@
+# credo:disable-for-this-file Credo.Check.Refactor.Nesting
+with {:module, StreamData} <- Code.ensure_loaded(StreamData) do
+  defmodule Purl.Generator do
+    @moduledoc """
+    `StreamData` generator for valid purls
+    """
+
+    import StreamData
+
+    alias Purl.Generator.Version, as: VersionGenerator
+
+    @spec type :: StreamData.t(Purl.type())
+    def type, do: require_letter_start(string([?a..?z, ?A..?Z, ?0..?9, ?., ?+, ?-], min_length: 1, max_length: 20))
+
+    @spec namespace_segment :: StreamData.t(Purl.namespace_segment())
+    def namespace_segment, do: filter(string(:printable, min_length: 1, max_length: 50), &(not String.contains?(&1, "/")))
+
+    @spec namespace :: StreamData.t(Purl.namespace())
+    def namespace, do: list_of(namespace_segment(), max_length: 10)
+
+    @spec name :: StreamData.t(Purl.name())
+    def name, do: string(:printable, min_length: 1, max_length: 50)
```

Comment:
Is `:printable` ASCII? Some package types allow non-ASCII characters: [`pkg:nuget/fels%C3%B6kning.utilities.svenskautilities@3.0.0`](https://www.nuget.org/packages/Fels%C3%B6kning.Utilities.SvenskaUtilities)

Reply:
Printable includes most of UTF8 including emojis etc. I’m not including all of UTF8 since that’s hard to debug when printing to the console.